test/test_cuda_sanitizer.py

```diff
@@ -0,0 +1,433 @@
+
+    def test_branch_sync(self):
+        # Tests that two streams can read after both waiting for a third, but they
+        # cannot write without further synchronization.
+
+        self.assert_good_kernel_launch(stream_id(1), read_write=[tensor_id(1)])
+        self.handler._handle_event_record(event_id(0), stream_id(1))
+        self.handler._handle_event_wait(event_id(0), stream_id(2))
+        self.handler._handle_event_wait(event_id(0), stream_id(3))
+        self.assert_good_kernel_launch(stream_id(2), read_only=[tensor_id(1)])
+        self.assert_good_kernel_launch(stream_id(3), read_only=[tensor_id(1)])
+
+        self.assert_bad_kernel_launch(stream_id(2), read_write=[tensor_id(1)])
+
+    def test_chain_sync(self):
+        iterations = 10
+
+        self.assert_good_kernel_launch(stream_id(0), read_only=[tensor_id(1)])
+        for i in range(iterations):
+            self.handler._handle_event_record(event_id(i), stream_id(i))
+            self.handler._handle_event_wait(event_id(i), stream_id(i + 1))
+        self.assert_good_kernel_launch(stream_id(iterations), read_write=[tensor_id(1)])
+
+    def test_expired_record(self):
+        self.assert_good_kernel_launch(stream_id(1), read_only=[tensor_id(1)])
+        self.handler._handle_event_record(event_id(0), stream_id(1))
+        self.assert_good_kernel_launch(stream_id(1), read_only=[tensor_id(1)])
+        self.handler._handle_event_wait(event_id(0), stream_id(2))
+
+        self.assert_bad_kernel_launch(stream_id(2), read_write=[tensor_id(1)])
+
+    def test_deleted_record(self):
+        for should_delete, should_create in [
+            (True, True),
+            (True, False),
+            (False, True),
+        ]:
+            self.setUp()
+            with self.subTest(should_delete=should_delete, should_create=should_create):
+                self.assert_good_kernel_launch(stream_id(1), read_only=[tensor_id(1)])
+                self.handler._handle_event_record(event_id(0), stream_id(1))
+
+                if should_delete:
+                    self.handler._handle_event_deletion(event_id(0))
+                if should_create:
+                    self.handler._handle_event_creation(event_id(0))
+
+                self.handler._handle_event_wait(event_id(0), stream_id(2))
+                self.assert_bad_kernel_launch(stream_id(2), read_write=[tensor_id(1)])
+
+    def test_all_reads_checked_failing(self):
+        iterations = 10
+        for i in range(1, iterations + 1):
+            self.assert_good_kernel_launch(stream_id(i), read_only=[tensor_id(1)])
+            self.handler._handle_event_record(event_id(i), stream_id(i))
+
+        for i in range(1, iterations):
+            self.handler._handle_event_wait(event_id(i), stream_id(0))
+
+        self.assert_bad_kernel_launch(stream_id(0), read_write=[tensor_id(1)])
+
+    def test_all_reads_checked_passing(self):
+        iterations = 10
+        for i in range(1, iterations):
```

Comment:
took me a few moments to notice difference between this and previous test, could you add a comment why the previous one is failing

Reply:
Done.